Adding Helm-To-Operator-Codegen-Sdk Documentation (#117)
nephio-project/docs: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md

---

> title: Helm to Operator Codegen Sdk Developer Guide

Reviewer: Is this a document for the developers of Helm to Operator Codegen Sdk?
If yes, then it should be moved to the Nephio contributor guides section.

Author: Yes, it is intended for developers, I will surely move it to  "Nephio contributor guides section."

weight: 1
---

The [Helm to Operator Codegen SDK](https://github.com/nephio-project/nephio-sdk/tree/main/helm-to-operator-codegen-sdk) offers a streamlined solution for translating existing Helm charts into Kubernetes operators with minimal effort and cost.

## The Flow Diagram

> In a nutshell, Firstly, the Helm-Charts are converted to Yamls using the values provided in "values.yaml". Then, each Kubernetes Resource Manifest (KRM) in the YAML is translated into Go code, employing one of two methods.

Reviewer 2: https://cloud.google.com/blog/topics/developers-practitioners/build-platform-krm-part-2-how-kubernetes-resource-model-works
```suggestion
In a nutshell, the Helm-Charts are converted to YAML files using the values provided in "values.yaml". Then, each Kubernetes Resource Model (KRM) in the YAML is translated into Go code, employing one of two methods.
```

1) If the resource is Runtime-Supported, it undergoes a conversion process where the KRM resource is first transformed into a Runtime Object, then into JSON, and finally into Go code.
2) Otherwise, if the resource is not Runtime-Supported, it is converted into an Unstructured Object and then into Go code.

> After the conversion process, all the generated Go code is gathered and compiled into a single Go file. This resulting file contains functions that can be readily utilized by Kubernetes operators.

Reviewer 2: ```suggestion
After the conversion process, all the generated Go code is gathered and compiled into a single Go file. This process results in a file that contains functions which can be readily utilized by Kubernetes operators.
```

![alt Flow Diagram](/static/images/user-guides/helm-to-operator-codegen-sdk-flow-diagram.jpg)

-----

> ### Flow-1: Helm to Yaml
> Helm to Yaml conversion is achieved by running the command

Reviewer 2: ```suggestion
### Flow-1: Helm to YAML
Helm to YAML conversion is achieved by running the following command
```

> `helm template <chart>  --namespace <namespace>  --output-dir “temp/templated/”` Internally. As of now, It retrieves the values from default "values.yaml"

Reviewer 2: ```suggestion
`helm template <chart>  --namespace <namespace>  --output-dir “temp/templated/”` Internally. As of now, it retrieves the values from default "values.yaml"
```

> ### Flow-2: Yaml Split

Reviewer 2: ```suggestion
### Flow-2: YAML Split
```

> The SDK iterates over each YAML file in the "converted-yamls" directory. If a YAML file contains multiple Kubernetes Resource Manifests (KRM), separated by "---", the SDK splits the YAML file accordingly to isolate each individual KRM resource. This ensures that each KRM resource is processed independently.

Reviewer 2: ```suggestion
The SDK iterates over each YAML file in the "converted-yamls" directory. If a YAML file contains multiple Kubernetes Resource Models (KRMs), separated by "---", the SDK splits the YAML file accordingly to isolate each individual KRM resource. This ensures that each KRM resource is processed independently.
```

			"metadata": map[string]interface{}{
				"name": "macvlan-conf",
			},
			"spec": map[string]interface{}{
				"config": "some-config",
			},
		},
	}
```
</details>

### Flow-3.1: KRM to Runtime-Object
The conversion process relies on the "k8s.io/apimachinery/pkg/runtime" package. Currently, only the API version "v1" is supported. The supported kinds for the Runtime Object method include:
`Deployment, Service, Secret, Role, RoleBinding, ClusterRoleBinding, PersistentVolumeClaim, StatefulSet, ServiceAccount, ClusterRole, PriorityClass, ConfigMap`

### Flow-3.2: Runtime-Object to JSON
Firstly, the SDK performs a typecast of the runtime object to its actual data type. For instance, if the Kubernetes Kind is "Service," the SDK typecasts the runtime object to the specific data type corev1.Service. Then, it conducts a Depth-First Search (DFS) traversal over the corev1.Service object using reflection. During this traversal, the SDK generates a JSON structure that encapsulates information about the struct hierarchy, including corresponding data types and values. This transformation results in a JSON representation of the corev1.Service object's structure and content.

> <details>

Reviewer: I would use the `<details>` tag here, It makes the content collapsed by default, so people will just not recognise it. Is this something what is not important?

Reviewer: Yes, exactly :)

        }
    },
    "Spec": {
        "type": "v1.ServiceSpec",
        "val": {
            "Ports": {
                "type": "[]v1.ServicePort",
                "val": [
                    {
                        "Port": {
                            "type": "int32",
                            "val": "80"
                        },
                        "Protocol": {
                            "type": "v1.Protocol",
                            "val": "TCP"
                        }
                    }
                ]
            }
        }
    }
}
// It shows the hierarchical structure along with the specific data types and corresponding values for each attribute
```
</details>

### Flow-3.3: JSON to String (Go-Code)
The SDK reads the JSON file containing the information about the Kubernetes resource and then translates this information into a string of Go code. This process involves parsing the JSON structure and generating corresponding Go code strings based on the structure, data types, and values extracted from the JSON representation. Ultimately, this results in a string that represents the Kubernetes resource in a format compatible with Go code.

> <details>

Reviewer: I think this should be simply just normal description.

<summary>TraverseJSON Cases (Json-to-String)</summary>

> ```

Reviewer: Why is this intro text is added as a code example?

                "val": [
                    {
                        "Port": {
                            "type": "int32",
                            "val": "80"
                        },
                        "Protocol": {
                            "type": "v1.Protocol",
                            "val": "TCP"
                        }
                    }
                ]
            }
        }
    }
}

// The Go-code will look like as
 &corev1.Service{
	ObjectMeta: metav1.ObjectMeta{
		Name: "my-service",
	},
	
	Spec: corev1.ServiceSpec{
		Ports: []corev1.ServicePort{
	
			corev1.ServicePort{
				Port:     80,
				Protocol: corev1.Protocol("TCP"),
				TargetPort: intstr.IntOrString{
					IntVal: 9376,
				},
			},
		},
	},
}
```
</details>

### Significance of Config-Jsons: (Struct_Module_mapping.json & Enum_module_mapping.json)
Based on the data type, Values are formatted accordingly,
| Data-Type | Value    | Formatted-Value    |
| :---:   | :---: | :---: |
| int32 | 5   | 5   |
| string | 5   | \"5\" |
| *int32 | 5   | int32Ptr(5) |

The Config-Jsons are required for more package-specific-types (such as : v1.Service, v1.Deployment)

> #### i) Struct_Module_mapping.json

Reviewer: ```suggestion
#### Struct_Module_mapping.json
```

Mostly, It is seen that inspecting the type of struct(using reflect) would tell us that the struct belong to package “v1”, but there are multiple v1 packages (appsv1, metav1, rbacv1, etc), So, the actual package remains unknown. 

Solution: To solve the above problems, we build a “structModuleMapping” which is a map that takes “struct-name” as the key and gives “package/module name” as a value.
```
v1.Deployment -->  appsv1.Deployment
v1.Service --> corev1.Service
```

> #### ii) Enum_Module_mapping.json

Reviewer: ```suggestion
#### Enum_Module_mapping.json
```

Structs need to be initialized using curly brackets {}, whereas enums need Parenthesis (), Since, reflect doesn’t tell us which data-type is struct or enum, We:

Solution: We solve the above problems by building an “enumModuleMapping” which is a set that stores all data types that are enums. i.e. If a data type belongs to the set, then It is an Enum.

There is an automation-script that takes the types.go files of packages and build the config-json. For details, Please refer [here](https://github.com/nephio-project/nephio-sdk/tree/main/helm-to-operator-codegen-sdk/config)

### Flow-4: KRM to Unstruct-Obj to String(Go-code)
All Kubernetes resource kinds that are not supported by the runtime-object method are handled using the unstructured method. In this approach, the Kubernetes Resource MOdel (KRM) is converted to an unstructured object using the package "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured". 
Then, We traverse the unstructured-Obj in a DFS fashion and build the gocode-string.

> <details>

Reviewer: I would leave the `<details>`

<summary>DFS Algorithm Cases (Unstruct-Version)</summary>

> ```	

Reviewer: All of this is normal description and not example code.
